pypeit/edgetrace.py

```diff
@@ -0,0 +1,4109 @@
+            self.fit_refine(debug=debug)
+            # Use the fits to determine if there are any discontinous
+            # trace centroid measurements that are actually components
+            # of the same slit edge
+            self.merge_traces(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        # Check if the PCA decomposition is possible; this should catch
+        # long slits
+        if self.can_pca():
+            # Use a PCA decomposition to parameterize the trace
+            # functional forms
+            self.pca_refine(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+            # Use the results of the PCA decomposition to recify and
+            # detect peaks/troughs in the spectrally collapsed
+            # Sobel-filtered image, then use those peaks to further
+            # refine the edge traces
+            self.peak_refine(rebuild_pca=True, debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        elif self.par['sync_predict'] == 'pca':
+            # TODO: Maybe we should instead force a fault here instead
+            # of letting the code change a parameter?
+            msgs.warn('Sync predict cannot use PCA; using nearest instead.')
+            self.par['sync_predict'] = 'nearest'
+
+            # NOTE: If the PCA decomposition is possible, the
+            # subsequent call to trace.peak_trace (called by
+            # peak_refine) removes all the existing traces and replaces
+            # them with traces at the peak locations. Those traces are
+            # then fit, regardless of the length of the value centroid
+            # measurements. So coming out of peak_refine there are no
+            # traces that are fully masked, which is not true if that
+            # block of code isn't run. That means for sync to work
+            # correctly, we have to remove fully masked traces. This is
+            # done inside sync().
+
+        # Left-right synchronize the traces
+        self.sync()
+        if show_stages:
+            self.show(thin=10, include_img=True, idlabel=True)
+
+        # TODO: Add a parameter and an if statement that will allow for
+        # this.
+        # `peak_refine` ends with the traces being described by a
+        # polynomial. Instead finish by reconstructing the trace models
+        # using the PCA decomposition
+#        self.pca_refine(debug=debug)
+#        if show_stages:
+#            self.show(thin=10, include_img=True, idlabel=True)
+            
+        # TODO: Add mask_refine() when it's ready
+
+        # Add this to the log
+        self.log += [inspect.stack()[0][3]]
+        if save:
+            # Save the object to a file
+            self.save()
+
+    def initial_trace(self, img, bpm=None, det=1, binning='1,1', save=False):
```

Comment:
It's only optional to allow for `img` to be a `numpy.ndarray` instead of a `TraceImage` object.  If `img` is a `TraceImage` object, then it just pulls the binning from that object.